source/Octopus.Tentacle/Services/Scripts/ScriptServiceV2.cs

```diff
@@ -0,0 +1,163 @@
+
+        public ScriptServiceV2(
+            IShell shell,
+            IScriptWorkspaceFactory workspaceFactory,
+            IScriptStateStoreFactory scriptStateStoreFactory,
+            ISystemLog log)
+        {
+            this.shell = shell;
+            this.workspaceFactory = workspaceFactory;
+            this.scriptStateStoreFactory = scriptStateStoreFactory;
+            this.log = log;
+        }
+
+        public ScriptStatusResponseV2 StartScript(StartScriptCommandV2 command)
+        {
+            var runningScript = runningScripts.GetOrAdd(
+                command.ScriptTicket,
+                s =>
+                {
+                    var workspace = workspaceFactory.GetWorkspace(command.ScriptTicket);
+                    var scriptState = scriptStateStoreFactory.Get(s, workspace);
+                    return new RunningScriptWrapper(scriptState);
+                });
+
+            using (runningScript.ScriptStateStore.GetExclusiveLock())
+            {
+                IScriptWorkspace workspace;
+
+                // If the state already exists then this runningScript is already running/has already run and we should not run it again
+                if (runningScript.ScriptStateStore.Exists())
+                {
+                    var state = runningScript.ScriptStateStore.Load();
+
+                    if (state.HasStarted() || runningScript.Process != null)
+                    {
+                        return GetResponse(command.ScriptTicket, 0, runningScript.Process);
+                    }
+
+                    workspace = workspaceFactory.GetWorkspace(command.ScriptTicket);
+                }
+                else
+                {
+                    workspace = workspaceFactory.PrepareWorkspace(command.ScriptTicket,
+                        command.ScriptBody,
+                        command.Scripts,
+                        command.Isolation,
+                        command.ScriptIsolationMutexTimeout,
+                        command.IsolationMutexName,
+                        command.Arguments,
+                        command.Files);
+
+                    runningScript.ScriptStateStore.Create();
+                }
+
+                var cancellationTokenSource = new CancellationTokenSource();
+                var process = LaunchShell(command.ScriptTicket, command.TaskId, workspace, runningScript.ScriptStateStore, cancellationTokenSource);
+
+                runningScript.Process = process;
+                runningScript.CancellationTokenSource = cancellationTokenSource;
+
+                if (command.DurationToWaitForScriptToFinish > TimeSpan.Zero)
+                {
+                    var waited = Stopwatch.StartNew();
+                    while (process.State != ProcessState.Complete && waited.Elapsed < command.DurationToWaitForScriptToFinish)
+                    {
+                        Thread.Sleep(TimeSpan.FromMilliseconds(100));
```

Comment:
This could be dropped even lower, I would go 10ms or less. The amount of CPU cycles it takes to wake up is going to be way less than the halibut request responses sent if this new feature was not used.